WebServices/AzureSignalR/ChatClient/SignalRService.cs

```diff
@@ -0,0 +1,101 @@
+﻿using ChatClient.Model;
+using Microsoft.AspNetCore.SignalR.Client;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ChatClient
+{
+    public class SignalRService
+    {
+        HttpClient client;
+
+        public delegate void MessageReceivedHandler(object sender, Message message);
+        public delegate void ConnectionHandler(object sender, bool successful, string message);
+
+        public event MessageReceivedHandler NewMessageReceived;
+        public event ConnectionHandler Connected;
+        public event ConnectionHandler ConnectionFailed;
+        public bool IsConnected { get; private set; }
+        public bool IsBusy { get; private set; }
+
+        public SignalRService()
+        {
+            client = new HttpClient();
+        }
+
+        public async Task SendMessageAsync(string username, string message)
+        {
+            IsBusy = true;
+
+            var newMessage = new Message
+            {
+                Name = username,
+                Text = message
+            };
+
+            var json = JsonConvert.SerializeObject(newMessage);
+            var content = new StringContent(json, Encoding.UTF8, "application/json");
+            var result = await client.PostAsync($"{Constants.HostName}/api/talk", content);
+
+            IsBusy = false;
+        }
+
+        public async Task ConnectAsync()
+        {
+            try
+            {
+                IsBusy = true;
+
+                string negotiateJson = await client.GetStringAsync($"{Constants.HostName}/api/negotiate");
+                NegotiateInfo negotiate = JsonConvert.DeserializeObject<NegotiateInfo>(negotiateJson);
+                HubConnection connection = new HubConnectionBuilder()
+                    .WithUrl(negotiate.Url, options =>
+                    {
+                        options.AccessTokenProvider = async () => negotiate.AccessToken;
```

Comment:
Is `async` needed here?

Reply:
Debatable: `AccessTokenProvider` expects a function that returns a `Task<string>`. There are other ways to do that but this seems to be the simplest way to wrap the `AccessToken` string in a Task. If you have a cleaner idea let me know!